File: vox/rust/vox-core/src/transport_prologue.rs
use vox_types::{Link, LinkRx, LinkTx, SplitLink};
use zerocopy::FromBytes;
use zerocopy::little_endian::U32 as LeU32;

const TRANSPORT_HELLO_MAGIC: u32 = u32::from_le_bytes(*b"VOTH");
const TRANSPORT_ACCEPT_MAGIC: u32 = u32::from_le_bytes(*b"VOTA");
const TRANSPORT_REJECT_MAGIC: u32 = u32::from_le_bytes(*b"VOTR");
const TRANSPORT_VERSION: u8 = 9;
const REJECT_UNSUPPORTED_PROLOGUE: u8 = 1;
const RESERVED_ZERO: [u8; 3] = [0; 3];

#[derive(Debug)]
pub enum TransportPrologueError {
    Io(std::io::Error),
    LinkDead,
    Protocol(String),
    Rejected(TransportRejectReason),
}

impl std::fmt::Display for TransportPrologueError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Io(error) => write!(f, "io error: {error}"),
            Self::LinkDead => write!(f, "link closed during transport prologue"),
            Self::Protocol(message) => write!(f, "protocol error: {message}"),
            Self::Rejected(reason) => write!(f, "transport rejected: {reason}"),
        }
    }
}

impl std::error::Error for TransportPrologueError {}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TransportRejectReason {
    UnsupportedPrologue,
}

impl std::fmt::Display for TransportRejectReason {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::UnsupportedPrologue => write!(f, "unsupported transport prologue"),
        }
    }
}

#[derive(
    Clone,
    Copy,
    zerocopy::FromBytes,
    zerocopy::IntoBytes,
    zerocopy::KnownLayout,
    zerocopy::Immutable,
)]
#[repr(C)]
struct TransportHello {
    magic: LeU32,
    version: u8,
    reserved: [u8; 3],
}

#[derive(
    Clone,
    Copy,
    zerocopy::FromBytes,
    zerocopy::IntoBytes,
    zerocopy::KnownLayout,
    zerocopy::Immutable,
)]
#[repr(C)]
struct TransportAccept {
    magic: LeU32,
    version: u8,
    reserved: [u8; 3],
}

#[derive(
    Clone,
    Copy,
    zerocopy::FromBytes,
    zerocopy::IntoBytes,
    zerocopy::KnownLayout,
    zerocopy::Immutable,
)]
#[repr(C)]
struct TransportReject {
    magic: LeU32,
    version: u8,
    reason: u8,
    reserved: [u8; 2],
}

// r[impl transport.prologue]
// r[impl transport.prologue.request]
// r[impl transport.prologue.accept]
// r[impl transport.prologue.reject-close]
pub async fn initiate_transport<L: Link>(
    link: L,
) -> Result<SplitLink<L::Tx, L::Rx>, TransportPrologueError> {
    let (tx, mut rx) = link.split();
    let hello = TransportHello {
        magic: LeU32::new(TRANSPORT_HELLO_MAGIC),
        version: TRANSPORT_VERSION,
        reserved: RESERVED_ZERO,
    };
    send_message(&tx, &hello).await?;

    let raw = recv_bytes(&mut rx).await?;
    let bytes = raw.as_bytes();
    let magic = bytes
        .get(..4)
        .and_then(|prefix| prefix.try_into().ok())
        .map(u32::from_le_bytes)
        .ok_or_else(|| {
            TransportPrologueError::Protocol("transport prologue message size mismatch".into())
        })?;

    if magic == TRANSPORT_ACCEPT_MAGIC {
        let accept = TransportAccept::read_from_bytes(bytes).map_err(|_| {
            TransportPrologueError::Protocol("transport prologue message size mismatch".into())
        })?;
        if accept.version != TRANSPORT_VERSION {
            return Err(TransportPrologueError::Protocol(format!(
                "unsupported transport version {}",
                accept.version
            )));
        }
        if accept.reserved != RESERVED_ZERO {
            return Err(TransportPrologueError::Protocol(
                "transport accept reserved bytes must be zero".into(),
            ));
        }
        return Ok(SplitLink { tx, rx });
    }

    if magic == TRANSPORT_REJECT_MAGIC {
        let reject = TransportReject::read_from_bytes(bytes).map_err(|_| {
            TransportPrologueError::Protocol("transport prologue message size mismatch".into())
        })?;
        if reject.version != TRANSPORT_VERSION {
            return Err(TransportPrologueError::Protocol(format!(
                "unsupported transport version {}",
                reject.version
            )));
        }
        let reason = match reject.reason {
            REJECT_UNSUPPORTED_PROLOGUE => TransportRejectReason::UnsupportedPrologue,
            other => {
                return Err(TransportPrologueError::Protocol(format!(
                    "unknown transport reject reason {other}"
                )));
            }
        };
        return Err(TransportPrologueError::Rejected(reason));
    }

    Err(TransportPrologueError::Protocol(
        "expected TransportAccept or TransportReject".into(),
    ))
}

// r[impl transport.prologue]
// r[impl transport.prologue.first-payload]
// r[impl transport.prologue.request]
// r[impl transport.prologue.accept]
// r[impl transport.prologue.reject-close]
pub async fn accept_transport<L: Link>(
    link: L,
) -> Result<SplitLink<L::Tx, L::Rx>, TransportPrologueError> {
    let (tx, mut rx) = link.split();
    let hello = recv_message::<_, TransportHello>(&mut rx).await?;
    if hello.magic.get() != TRANSPORT_HELLO_MAGIC {
        return Err(TransportPrologueError::Protocol(
            "transport hello magic mismatch".into(),
        ));
    }
    if hello.version != TRANSPORT_VERSION {
        return Err(TransportPrologueError::Protocol(format!(
            "unsupported transport version {}",
            hello.version
        )));
    }
    if hello.reserved != RESERVED_ZERO {
        reject_transport(&tx, TransportRejectReason::UnsupportedPrologue).await?;
        return Err(TransportPrologueError::Protocol(
            "transport hello reserved bytes must be zero".into(),
        ));
    }

    let accept = TransportAccept {
        magic: LeU32::new(TRANSPORT_ACCEPT_MAGIC),
        version: TRANSPORT_VERSION,
        reserved: RESERVED_ZERO,
    };
    send_message(&tx, &accept).await?;
    Ok(SplitLink { tx, rx })
}

// r[impl transport.prologue.accept]
// r[impl transport.prologue.reject-close]
pub async fn reject_transport<L: LinkTx>(
    tx: &L,
    reason: TransportRejectReason,
) -> Result<(), TransportPrologueError> {
    let code = match reason {
        TransportRejectReason::UnsupportedPrologue => REJECT_UNSUPPORTED_PROLOGUE,
    };
    let reject = TransportReject {
        magic: LeU32::new(TRANSPORT_REJECT_MAGIC),
        version: TRANSPORT_VERSION,
        reason: code,
        reserved: [0; 2],
    };
    send_message(tx, &reject).await
}

async fn send_message<LTx: LinkTx, M: zerocopy::IntoBytes + zerocopy::Immutable>(
    tx: &LTx,
    message: &M,
) -> Result<(), TransportPrologueError> {
    tx.send(message.as_bytes().to_vec())
        .await
        .map_err(TransportPrologueError::Io)
}

async fn recv_message<
    LRx: LinkRx,
    M: zerocopy::FromBytes + zerocopy::KnownLayout + zerocopy::Immutable,
>(
    rx: &mut LRx,
) -> Result<M, TransportPrologueError> {
    let raw = recv_bytes(rx).await?;
    M::read_from_bytes(raw.as_bytes()).map_err(|_| {
        TransportPrologueError::Protocol("transport prologue message size mismatch".into())
    })
}

async fn recv_bytes<LRx: LinkRx>(
    rx: &mut LRx,
) -> Result<vox_types::Backing, TransportPrologueError> {
    rx.recv()
        .await
        .map_err(|error| {
            TransportPrologueError::Protocol(format!("transport recv failed: {error}"))
        })?
        .ok_or(TransportPrologueError::LinkDead)
}

File: vox/typescript/packages/vox-core/src/transport_prologue.ts
import type { Link } from "./link.ts";

const TRANSPORT_HELLO_MAGIC = new Uint8Array([0x56, 0x4F, 0x54, 0x48]); // VOTH
const TRANSPORT_ACCEPT_MAGIC = new Uint8Array([0x56, 0x4F, 0x54, 0x41]); // VOTA
const TRANSPORT_REJECT_MAGIC = new Uint8Array([0x56, 0x4F, 0x54, 0x52]); // VOTR
const TRANSPORT_VERSION = 9;
const TRANSPORT_RESERVED_ZERO = new Uint8Array([0, 0, 0]);
const REJECT_UNSUPPORTED_PROLOGUE = 1;

export class TransportPrologueError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TransportPrologueError";
  }
}

function sameBytes(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) {
    return false;
  }
  for (let i = 0; i < left.length; i++) {
    if (left[i] !== right[i]) {
      return false;
    }
  }
  return true;
}

function encodeHello(): Uint8Array {
  return new Uint8Array([
    ...TRANSPORT_HELLO_MAGIC,
    TRANSPORT_VERSION,
    ...TRANSPORT_RESERVED_ZERO,
  ]);
}

function encodeAccept(): Uint8Array {
  return new Uint8Array([
    ...TRANSPORT_ACCEPT_MAGIC,
    TRANSPORT_VERSION,
    ...TRANSPORT_RESERVED_ZERO,
  ]);
}

function encodeReject(reason = REJECT_UNSUPPORTED_PROLOGUE): Uint8Array {
  return new Uint8Array([
    ...TRANSPORT_REJECT_MAGIC,
    TRANSPORT_VERSION,
    reason,
    0,
    0,
  ]);
}

function reservedBytesAreZero(bytes: Uint8Array): boolean {
  return sameBytes(bytes, TRANSPORT_RESERVED_ZERO);
}

// r[impl transport.prologue]
// r[impl transport.prologue.request]
// r[impl transport.prologue.accept]
// r[impl transport.prologue.reject-close]
export async function performInitiatorTransportPrologue(link: Link): Promise<void> {
  await link.send(encodeHello());
  const response = await link.recv();
  if (!response) {
    throw new TransportPrologueError("transport closed during prologue");
  }
  if (response.length !== 8) {
    throw new TransportPrologueError("invalid transport prologue response size");
  }
  if (sameBytes(response.subarray(0, 4), TRANSPORT_ACCEPT_MAGIC)) {
    if ((response[4] ?? 0) !== TRANSPORT_VERSION) {
      throw new TransportPrologueError(`unsupported transport version ${response[4] ?? 0}`);
    }
    if (!reservedBytesAreZero(response.subarray(5, 8))) {
      throw new TransportPrologueError("transport accept reserved bytes must be zero");
    }
    return;
  }
  if (sameBytes(response.subarray(0, 4), TRANSPORT_REJECT_MAGIC)) {
    const reason = response[5] ?? 0;
    if (reason === REJECT_UNSUPPORTED_PROLOGUE) {
      throw new TransportPrologueError("transport rejected unsupported prologue");
    }
    throw new TransportPrologueError(`transport rejected with reason ${reason}`);
  }
  throw new TransportPrologueError("expected TransportAccept or TransportReject");
}

// r[impl transport.prologue]
// r[impl transport.prologue.first-payload]
// r[impl transport.prologue.request]
// r[impl transport.prologue.accept]
// r[impl transport.prologue.reject-close]
export async function performAcceptorTransportPrologue(link: Link): Promise<void> {
  const hello = await link.recv();
  if (!hello) {
    throw new TransportPrologueError("transport closed before prologue");
  }
  if (hello.length !== 8 || !sameBytes(hello.subarray(0, 4), TRANSPORT_HELLO_MAGIC)) {
    throw new TransportPrologueError("invalid TransportHello");
  }
  if ((hello[4] ?? 0) !== TRANSPORT_VERSION) {
    await link.send(encodeReject());
    throw new TransportPrologueError(`unsupported transport version ${hello[4] ?? 0}`);
  }
  if (!reservedBytesAreZero(hello.subarray(5, 8))) {
    await link.send(encodeReject());
    throw new TransportPrologueError("transport hello reserved bytes must be zero");
  }
  await link.send(encodeAccept());
}
